streams/src/test/java/org/apache/kafka/streams/integration/HighAvailabilityTaskAssignorIntegrationTest.java

```
@@ -0,0 +1,266 @@
+                mkEntry(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName())
+            )
+        );
+
+
+        try (final KafkaStreams kafkaStreams0 = new KafkaStreams(topology, streamsProperties(appId, assignmentListener));
+             final KafkaStreams kafkaStreams1 = new KafkaStreams(topology, streamsProperties(appId, assignmentListener));
+             final Consumer<String, String> consumer = new KafkaConsumer<>(consumerProperties)) {
+            kafkaStreams0.start();
+
+            // wait until all the input records are in the changelog
+            TestUtils.waitForCondition(
+                () -> getChangelogOffsetSum(changelogTopicPartitions, consumer) == 1000,
+                120_000L,
+                () -> "Input records haven't all been written to the changelog: " + getChangelogOffsetSum(changelogTopicPartitions, consumer)
+            );
+
+            final AtomicLong instance1TotalRestored = new AtomicLong(-1);
+            final AtomicLong instance1NumRestored = new AtomicLong(-1);
+            final CountDownLatch restoreCompleteLatch = new CountDownLatch(1);
+            kafkaStreams1.setGlobalStateRestoreListener(new StateRestoreListener() {
+                @Override
+                public void onRestoreStart(final TopicPartition topicPartition,
+                                           final String storeName,
+                                           final long startingOffset,
+                                           final long endingOffset) {
+                }
+
+                @Override
+                public void onBatchRestored(final TopicPartition topicPartition,
+                                            final String storeName,
+                                            final long batchEndOffset,
+                                            final long numRestored) {
+                    instance1NumRestored.accumulateAndGet(
```

This should never be called -- would it be simpler if the just throw an exception if and of the listener methods is called?

---

I'm not sure; that seems to make assumptions about how exactly the restore logic would behave when you're already caught up.

I guess the same reasoning would apply to the other callback, we should never call it with a nonzero number of restored records, so we could just throw an exception from the callback.

In both cases, though, it seems better to make assertions in the scope of the test, rather than deep inside a callback in one of the threads in one of the streams instances and hope that it somehow bubbles up into a test failure.